src/stan/mcmc/hmc/nuts/base_nuts.hpp

```diff
@@ -70,83 +54,69 @@ namespace stan {
         ps_point z_sample(z_plus);
         ps_point z_propose(z_plus);
 
-        int n_cont = init_sample.cont_params().size();
-
-        Eigen::VectorXd rho_init = this->z_.p;
-        Eigen::VectorXd rho_plus(n_cont); rho_plus.setZero();
-        Eigen::VectorXd rho_minus(n_cont); rho_minus.setZero();
-
-        util.H0 = this->hamiltonian_.H(this->z_);
+        Eigen::VectorXd rho = this->z_.p;
+        double sum_weight = 1;
 
-        // Sample the slice variable
-        util.log_u = std::log(this->rand_uniform_());
-
-        // Build a balanced binary tree until the NUTS criterion fails
-        util.criterion = true;
-        int n_valid = 0;
+        double H0 = this->hamiltonian_.H(this->z_);
+        int n_leapfrog = 0;
+        double sum_metro_prob = 1;  // exp(H0 - H0)
 
+        // Build a trajectory until the NUTS criterion is no longer satisfied
         this->depth_ = 0;
         this->divergent_ = 0;
 
-        util.n_tree = 0;
-        util.sum_prob = 0;
+        while (this->depth_ < this->max_depth_) {
+          // Build a new subtree in a random direction
+          Eigen::VectorXd rho_subtree(rho.size());
+          rho_subtree.setZero();
 
-        while (util.criterion && (this->depth_ <= this->max_depth_)) {
-          // Randomly sample a direction in time
-          ps_point* z = 0;
-          Eigen::VectorXd* rho = 0;
+          bool valid_subtree = false;
+          double sum_weight_subtree = 0;
 
           if (this->rand_uniform_() > 0.5) {
-            z = &z_plus;
-            rho = &rho_plus;
-            util.sign = 1;
+            this->z_.ps_point::operator=(z_plus);
```

Comment:
This copies only the position and momenta, not the mass matrix.  It avoids unnecessary copies which was a significant burden.

> On Apr 8, 2016, at 17:01, Bob Carpenter <notifications@github.com> wrote:
> 
> In src/stan/mcmc/hmc/nuts/base_nuts.hpp:
> 
> >  
> >            if (this->rand_uniform_() > 0.5) {
> > -            z = &z_plus;
> > -            rho = &rho_plus;
> > -            util.sign = 1;
> > +            this->z_.ps_point::operator=(z_plus);
> Why not use the operator syntax?
> 
> this->z_ = z_plus;
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
>